### PR TITLE
Convert compressor operation from int constexpr to enum

### PR DESCRIPTION
**Pull request by jmythms**
This PR attempts to convert compressor operation in multiple files specified by int constexpr to enum class. 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataHVACGlobals.hh [hunk 1]
```diff
@@ -366,6 +366,14 @@ namespace DataHVACGlobals {
...
     };
 
+    // Compressor operation
+    enum class CompressorOperation
+    {
+        Unassigned = -1,
+        Off, // signal DXCoil that compressor shouldn't run
+        On,  // normal compressor operation
+        Num
+    };
```
Comment (jmythms) [inline src/EnergyPlus/DataHVACGlobals.hh]: `CompressorOperation` enum class exists in DataHVACGlobals.hh.

File: src/EnergyPlus/Furnaces.cc [hunk 2]
```diff
@@ -9794,7 +9869,7 @@ namespace Furnaces {
             FirstHVACIteration = false;
         }
         FanOpMode = int(Par(3));
-        CompOp = int(Par(4));
+        CompOp = static_cast<CompressorOperation>(Par(4));
```
Comment (jmythms) [inline src/EnergyPlus/Furnaces.cc]: The `par(*)` values that are parameters and results of the residual functions had to be `static_cast`.

File: src/EnergyPlus/IntegratedHeatPump.hh [hunk 3]
```diff
@@ -54,6 +54,7 @@
 
 // EnergyPlus Headers
 #include <EnergyPlus/Data/BaseData.hh>
+#include <EnergyPlus/DataHVACGlobals.hh>
```
Comment (jmythms) [inline src/EnergyPlus/IntegratedHeatPump.hh:57]: Had to include the DataHVACGlobals.hh here.

File: src/EnergyPlus/PackagedTerminalHeatPump.cc [hunk 4]
```diff
@@ -6273,8 +6273,8 @@ void CalcPTUnit(EnergyPlusData &state,
     // Tuned Named constants to avoid heap allocation when passed to Optional args
     bool constexpr True(true);
     bool constexpr False(false);
-    int constexpr iZero(0);
-    int constexpr iOne(1);
+    CompressorOperation constexpr iZero(CompressorOperation::Off);
+    CompressorOperation constexpr iOne(CompressorOperation::On);
```
Comment (jmythms) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc]: Will refactor 😅

File: src/EnergyPlus/UnitarySystem.cc [hunk 5]
```diff
@@ -233,7 +231,7 @@ namespace UnitarySystems {
                                  Real64 &sysOutputProvided,
                                  Real64 &latOutputProvided)
     {
-        int CompOn = 0;
+        DataHVACGlobals::CompressorOperation CompOn = DataHVACGlobals::CompressorOperation::Off;
```
Comment (jmythms) [inline src/EnergyPlus/UnitarySystem.cc:234]: A `using namespace DataHVACGlobals;` statement will shorten these lines, make them more readable. Is that recommended?

File: src/EnergyPlus/UnitarySystem.cc [hunk 6]
```diff
@@ -11200,28 +11198,49 @@ namespace UnitarySystems {
 
                 DXCoils::SimDXCoil(
                     state, blankString, CompOn, FirstHVACIteration, CompIndex, this->m_FanOpMode, PartLoadRatio, OnOffAirFlowRatio, CoilCoolHeatRat);
-                this->m_CoolCompPartLoadRatio = PartLoadRatio * double(CompOn);
+                if (CompOn == DataHVACGlobals::CompressorOperation::On) {
+                    this->m_CoolCompPartLoadRatio = PartLoadRatio;
+                } else {
+                    this->m_CoolCompPartLoadRatio = 0;
+                }
```
Comment (jmythms) [inline src/EnergyPlus/UnitarySystem.cc]: Unrolled these multiplying with `double(CompOn)` to if else statements checking if the Compressor is on or of.

File: src/EnergyPlus/UnitarySystem.hh [hunk 7]
```diff
@@ -54,6 +54,7 @@
 
 // EnergyPlus headers
 #include <EnergyPlus/Data/BaseData.hh>
+#include <EnergyPlus/DataHVACGlobals.hh>
```
Comment (jmythms) [inline src/EnergyPlus/UnitarySystem.hh:57]: Another `#include <EnergyPlus/DataHVACGlobals.hh>` statement addition here.

File: src/EnergyPlus/Furnaces.cc (hunk 2 above)
Comment (amirroth) [reply src/EnergyPlus/Furnaces.cc]: These things need to be converted to lambdas anyway. There is already an open [issue](https://github.com/NREL/EnergyPlus/issues/9110) about that. That conversion will make these casts unnecessary.

File: src/EnergyPlus/UnitarySystem.hh [hunk 8]
```diff
@@ -776,22 +777,22 @@ namespace UnitarySystems {
...
 
         void calcUnitaryCoolingSystem(EnergyPlusData &state,
-                                      int const AirLoopNum,          // index to air loop
-                                      bool const FirstHVACIteration, // True when first HVAC iteration
-                                      Real64 const PartLoadRatio,    // coil operating part-load ratio
-                                      int const CompOn,              // compressor control (0=off, 1=on)
+                                      int const AirLoopNum,                        // index to air loop
+                                      bool const FirstHVACIteration,               // True when first HVAC iteration
+                                      Real64 const PartLoadRatio,                  // coil operating part-load ratio
+                                      DataHVACGlobals::CompressorOperation CompOn, // compressor control (0=off, 1=on)
```
Comment (jmythms) [inline src/EnergyPlus/UnitarySystem.hh]: I removed const qualifiers in declarations wherever I saw them in the code I was changing. [Suggestion from clang-tidy](https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html).

File: src/EnergyPlus/DXCoils.cc [hunk 9]
```diff
@@ -131,9 +131,9 @@ using namespace ScheduleManager;
 // Functions
 
 void SimDXCoil(EnergyPlusData &state,
-               std::string_view CompName,     // name of the fan coil unit
-               int const CompOp,              // compressor operation; 1=on, 0=off
-               bool const FirstHVACIteration, // True when first HVAC iteration
+               std::string_view CompName,        // name of the fan coil unit
+               CompressorOperation const CompOp, // compressor operation; 1=on, 0=off
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc:135]: I hate to be pedantic (actually, I love it) but I think `ComprOp` or `CompressOp` or `CompressorOp` is better than `CompOp`. We are already using `Comp` to mean component (see preceding parameter), we shouldn't also use it to mean compressor.

File: src/EnergyPlus/DXCoils.cc [hunk 10]
```diff
@@ -229,19 +229,21 @@ void SimDXCoil(EnergyPlusData &state,
             //   call the HPWHDXCoil routine to calculate water side performance set up the DX coil info for air-side calcs
             CalcHPWHDXCoil(state, DXCoilNum, PartLoadRatio);
             //    CALL CalcDoe2DXCoil(state, DXCoilNum, CompOp, FirstHVACIteration,PartLoadRatio), perform air-side calculations
-            CalcDoe2DXCoil(state, DXCoilNum, 1, FirstHVACIteration, PartLoadRatio, FanOpMode);
+            CalcDoe2DXCoil(state, DXCoilNum, CompressorOperation::On, FirstHVACIteration, PartLoadRatio, FanOpMode);
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc:232]: I am not sure how many more of these kinds of things we have in EnergyPlus, but replacing integers with symbols is always a good thing. Make the code much more understandable and maintainable. This is a pattern we should propagate maximally.

File: src/EnergyPlus/DXCoils.cc (hunk 9 above)
Comment (jmythms) [reply src/EnergyPlus/DXCoils.cc:135]: Will change to `CompressorOp` 👍🏽

File: src/EnergyPlus/DXCoils.cc [hunk 11]
```diff
@@ -526,14 +528,14 @@ void SimDXCoilMultiMode(EnergyPlusData &state,
...
-                CalcDoe2DXCoil(state, DXCoilNum, On, FirstHVACIteration, PartLoadRatio, FanOpMode, PerfMode);
+                CalcDoe2DXCoil(state, DXCoilNum, CompressorOperation::On, FirstHVACIteration, PartLoadRatio, FanOpMode, PerfMode);
                 S1PLR = PartLoadRatio;
                 S2PLR = 0.0;
             } else {
                 // If a two-stage coil
                 // Run stage 1 at full load
                 PerfMode = DehumidMode * 2 + 1;
-                CalcDoe2DXCoil(state, DXCoilNum, On, FirstHVACIteration, 1.0, FanOpMode, PerfMode);
+                CalcDoe2DXCoil(state, DXCoilNum, CompressorOperation::On, FirstHVACIteration, 1.0, FanOpMode, PerfMode);
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc:538]: Not part of this PR, but I would also like to see some kind of pattern where non-symbolic arguments are documented inline, e.g.,: 

```cpp
CalcDoe2DXCoil(state, DXCoilNum, CompressorOperation::On, FirstHVACIteration, /* PartLoadRatio */1.0, FanOpMode, PerfMode);
```

Or 

```cpp
CalcDoe2DXCoil(state, DXCoilNum, CompressorOperation::On, FirstHVACIteration, 
                             1.0, // PartLoadRatio 
                             FanOpMode, PerfMode);
```

File: src/EnergyPlus/DXCoils.cc [hunk 12]
```diff
@@ -9534,7 +9536,7 @@ void CalcHPWHDXCoil(EnergyPlusData &state,
 
 void CalcDoe2DXCoil(EnergyPlusData &state,
                     int const DXCoilNum,                      // the number of the DX coil to be simulated
-                    int const CompOp,                         // compressor operation; 1=on, 0=off
+                    CompressorOperation const CompOp,         // compressor operation; 1=on, 0=off
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc]: Same comment about name `CompOp`.

File: src/EnergyPlus/DataHVACGlobals.hh [hunk 13]
```diff
@@ -366,6 +366,14 @@ namespace DataHVACGlobals {
         OptStartDataType() = default;
     };
 
+    // Compressor operation
+    enum class CompressorOperation
+    {
+        Unassigned = -1,
```
Comment (amirroth) [inline src/EnergyPlus/DataHVACGlobals.hh]: I think @mitchute is doing a pass that is converting `Unassigned` to `Invalid`.

Comment (jmythms) [reply src/EnergyPlus/DataHVACGlobals.hh]: I will change this one in this PR.

File: src/EnergyPlus/Furnaces.cc [hunk 14]
```diff
@@ -6203,12 +6226,34 @@ namespace Furnaces {
...
+                                                 LatentOutput,
+                                                 0.0,
+                                                 0.0,
+                                                 OnOffAirFlowRatio,
+                                                 SUPHEATERLOAD);
                         } else {
                             SetOnOffMassFlowRate(state, FurnaceNum, AirLoopNum, OnOffAirFlowRatio, OpMode, QZnReq, MoistureLoad, PartLoadRatio);
-                            CalcFurnaceOutput(
-                                state, FurnaceNum, false, 0, Off, 0.0, 0.0, 0.0, 0.0, SensibleOutput, LatentOutput, OnOffAirFlowRatio, false);
+                            CalcFurnaceOutput(state,
```
Comment (amirroth) [inline src/EnergyPlus/Furnaces.cc:6243]: Why is this function called `CalcFurnaceOutput` and not `CalcFurnace`?

File: src/EnergyPlus/SZVAVModel.cc [hunk 15]
```diff
@@ -977,7 +977,7 @@ namespace SZVAVModel {
                         bool const &HXUnitOn,
```
Comment (amirroth) [inline src/EnergyPlus/SZVAVModel.cc]: I thought we had gotten rid of these `bool const &` and `int const &` parameters.

File: src/EnergyPlus/UnitarySystem.cc (hunk 5 above)
Comment (amirroth) [reply src/EnergyPlus/UnitarySystem.cc:234]: I think we are trying to avoid `using` statements in general. The thing to do here is to shorten the name of the namespace itself. There should only be a dozen or so namespaces (there is no one-to-one relationship between namespace and file) and they should have names like `HVAC` and `Const` not `DataGlobalConstants` and `DataHVACGlobals`. You don't need long names when you have only 12 of something.

File: src/EnergyPlus/UnitarySystem.cc (hunk 6 above)
Comment (amirroth) [reply src/EnergyPlus/UnitarySystem.cc]: A shorthand for this idiom is:

```cpp
this->m_CoolCompPartLoadRatio = DataHVACGlobals::CompressorOperation::On ? PartLoadRatio : 0.0;
```

File: src/EnergyPlus/UnitarySystem.cc [hunk 16]
```diff
@@ -11200,28 +11198,49 @@ namespace UnitarySystems {
...
 
             } else if (SELECT_CASE_var == DataHVACGlobals::CoilDX_Cooling) { // CoilCoolingDX
 
                 bool const singleMode = (this->m_SingleMode == 1);
                 if (this->m_ControlType == ControlType::Setpoint) {
-                    if (this->m_CoolingSpeedNum > 1) {
```
Comment (amirroth) [inline src/EnergyPlus/UnitarySystem.cc:11209]: Yeah, this type of stuff (casting `enum` to `double`) is not good. Should scrub this idiom.

File: src/EnergyPlus/UnitarySystem.cc (hunk 5 above)
Comment (jmythms) [reply src/EnergyPlus/UnitarySystem.cc:234]: Got it. Should be a quick refactor when it happens.

File: src/EnergyPlus/UnitarySystem.cc (hunk 6 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: Um ... 

```cpp
this->m_CoolCompPartLoadRatio = CompOn ? PartLoadRatio : 0.0;
```

Comment (amirroth) [reply src/EnergyPlus/UnitarySystem.cc]: You caught me! Actually for clarity and type safety it should be:

```cpp
this->m_CoolCompPartLoadRatio = (compOp == DataHVACGlobals::CompressorOperation::On) ? PartLoadRatio : 0.0;
```

Comment (amirroth) [reply src/EnergyPlus/UnitarySystem.cc]: By the way, I made that mistake on purpose to see if anyone was paying attention! 😉

File: src/EnergyPlus/DXCoils.cc (hunk 11 above)
Comment (rraustad) [reply src/EnergyPlus/DXCoils.cc:538]: I like the first one, and never even thought of doing that.

File: src/EnergyPlus/UnitarySystem.cc (hunk 6 above)
Comment (jmythms) [reply src/EnergyPlus/UnitarySystem.cc]: Will make the changes, thanks for the suggestion! 😃

File: src/EnergyPlus/DXCoils.cc (hunk 11 above)
Comment (amirroth) [reply src/EnergyPlus/DXCoils.cc:538]: I like the first one also, but I think we are trying to avoid using C-style comments. @Myoldmopar, can you advise?

Comment (Myoldmopar) [reply src/EnergyPlus/DXCoils.cc:538]: Correct, we agreed to make C style comments illegal because they make parsing the code during refactors more difficult.  We have a CI script that will fail if you add C style arguments, so the C++ style will be the way to go.

File: src/EnergyPlus/UnitarySystem.cc (hunk 6 above)
Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: #9052 should easily merge with all these changes ...

File: src/EnergyPlus/Furnaces.cc (hunk 14 above)
Comment (jmythms) [reply src/EnergyPlus/Furnaces.cc:6243]: There is another function called `CalcFurnaceResidual` that calls `CalcFurnaceOutput`. Shall I rename `CalcFurnaceOutput` to `CalcFurnace`?

Comment (amirroth) [reply src/EnergyPlus/Furnaces.cc:6243]: Don't bother. I shouldn't have said anything.

Comment (jmythms) [reply src/EnergyPlus/Furnaces.cc:6243]: 😬

File: src/EnergyPlus/SimAirServingZones.hh [hunk 17]
```diff
@@ -66,10 +66,6 @@ struct EnergyPlusData;
 
 namespace SimAirServingZones {
 
-    // coil operation
-    constexpr int CoilOn(1);  // normal coil operation
-    constexpr int CoilOff(0); // signal coil shouldn't run
```
Comment (jmythms) [inline src/EnergyPlus/SimAirServingZones.hh:71]: These `CoilOn` and `CoilOff` variables were replaced with `CompressorOperation::On` and `CompressorOperation::Off` respectively. Reason: the argument of `SimHXAssistedCoolingCoil()` where they are used signals compressor operation.

**Comment by jmythms** [comment]
The comments have been addressed and this is ready for review.

**Comment by Myoldmopar** [comment]
This one next?  Seems like it is still ready for review, I'll pull develop and apply Clang Format here as well.

**Comment by Myoldmopar** [comment]
Super happy here, merging.

**Comment by jmythms** [comment]
Thank you! :)
